storey/sql_driver.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2022 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import List
+
+import sqlalchemy as db
+
+from storey.drivers import Driver
+
+
+class SQLDriver(Driver):
+    """
+    SQL database connector.
+    :param db_url: database url
+    :param primary_key: the primary key of the table.
+    """
+
+    def __init__(
+        self,
+        primary_key: str,
+        db_path: str,
+    ):
+        self._db_path = db_path
+        self._sql_connection = None
+        self._primary_key = primary_key
+
+    def _lazy_init(self):
+
+        if not self._sql_connection:
+            self._engine = db.create_engine(self._db_path)
+            self._sql_connection = self._engine.connect()
+
+    def _table(self, table_path):
+        import sqlalchemy as db
+
+        metadata = db.MetaData()
+        while table_path.startswith("/"):
+            table_path = table_path[1:]
+
+        return db.Table(
+            table_path.split("/")[1],
+            metadata,
+            autoload=True,
+            autoload_with=self._engine,
+        )
+
+    async def _save_schema(self, container, table_path, schema):
+        self._lazy_init()
+
+    async def _load_schema(self, container, table_path):
+        self._lazy_init()
+
+    async def _save_key(self, container, table_path, key, aggr_item, partitioned_by_key, additional_data):
+        self._lazy_init()
+
+        table = self._table(table_path)
+        try:
+            self._sql_connection.execute(table.insert(), [additional_data], autocommit=True)
+
+        except db.exc.IntegrityError:
+            raise RuntimeError(f"Failed to get insert {additional_data} to {table_path} table because IntegrityError")
+        await self.close()
+
+    async def _load_aggregates_by_key(self, container, table_path, key):
+        self._lazy_init()
+        table = self._table(table_path)
+
+        agg_val, values = await self._get_all_fields(key, table)
```

Comment:
How does this not fail? `_get_all_fields()` does not return a tuple. It must not be tested.